### ISSUE TITLE
Frequent redirects to not-found when switching between conversations

### ISSUE THREAD

**Issue by fritz-k** [nextcloud/spreed#4670]
Hi,

opening the same conversation again afterwards eventually leads to success. Roughly a quarter to a half of all requests get redirected to not-found.

## Steps to reproduce
1. Open an existing conversation

### Expected behaviour
The page shows the selected conversation.

### Actual behaviour
The page shows the selected conversation for a few seconds and then loads `index.php/apps/spreed/not-found` with the page stating that "The conversation does not exist".

## Talk app

**Talk app version:** 10.0.3

**Custom Signaling server configured:** no

**Custom TURN server configured:** no

**Custom STUN server configured:** no

## Browser

**Microphone available:** yes

**Camera available:** yes

**Operating system:** Linux; Ubuntu, Arch and NixOS

**Browser name:** Firefox & Chrom

**Browser version:** Various, including FF 82.0.2

### Browser log

<details>

```js
10:37:21.639 Current participant found CallButton.vue:101
10:37:22.293 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user4/44?v=3
10:37:22.321 XHRPOSThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v2/room/kjvse3k9/participants/active
[HTTP/2 200 OK 402ms]

10:37:22.468 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v1/chat/kjvse3k9?lookIntoFuture=0&lastKnownMessageId=10212&includeLastKnown=1
[HTTP/2 200 OK 343ms]

10:37:22.471 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/statuses/user4
[HTTP/2 200 OK 364ms]

10:37:22.481 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/collaboration/resources/room/kjvse3k9?format=json
[HTTP/2 200 OK 303ms]

10:37:22.870 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v1/signaling/settings?token=kjvse3k9
[HTTP/2 200 OK 603ms]

10:37:23.550 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user1/32
10:37:23.569 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user4/32?v=3
[HTTP/2 200 OK 0ms]

10:37:23.576 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user3/32
[HTTP/2 200 OK 0ms]

10:37:23.583 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user5/32
[HTTP/2 200 OK 0ms]

10:37:23.694 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user2/32
[HTTP/2 200 OK 0ms]

10:37:23.721 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user1/64
[HTTP/2 200 OK 0ms]

10:37:23.730 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user3/64
[HTTP/2 200 OK 0ms]

10:37:23.732 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user5/64
[HTTP/2 200 OK 0ms]

10:37:23.735 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user2/64
[HTTP/2 200 OK 0ms]

10:37:23.764 Joined signaling.js:192:10
10:37:23.896 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v1/chat/kjvse3k9?lookIntoFuture=1&lastKnownMessageId=10212&includeLastKnown=0
10:37:24.061 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v1/signaling/kjvse3k9
[HTTP/2 404 Not Found 403ms]

10:37:24.063 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v2/room/kjvse3k9/participants?includeStatus=true
[HTTP/2 200 OK 404ms]

10:37:24.512 Conversation was not found anymore talk.js:56:58520
    _startPullingMessages https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:56
    _startPullingMessages https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:56
    _joinRoomSuccess https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:56
    joinRoom https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:56
    joinRoom https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:56
    e https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:240
    c https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    _invoke https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    t https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    ue https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    (Async: promise callback)
    ue https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    le https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    le https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:220
    De https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:240
    Se https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:240
    e https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:437
    c https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    _invoke https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    t https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    g https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    (Async: promise callback)
    g https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    f https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    f https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    p https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:437
    onRouteChange https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2787
    Qe https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    $emit https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    n https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2850
    beforeMount https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2850
    p https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    r https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    r https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    pn https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    confirmTransition https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    transitionTo https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    push https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
10:37:24.531 Navigated to https://nextcloud.ourdomain.de/index.php/apps/spreed/not-found
10:37:24.531 TypeError: o.data.ocs is undefined
    n https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    c https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    _invoke https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    t https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    Rr https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    promise callback*Rr https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    Qr https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    Qr https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    cancelableGetParticipants https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    onJoinedConversation https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    tt https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    Ze https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    promise callback*qe https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    tt https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    $nextTick https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    onJoinedConversation https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2828
    Qe https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    $emit https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    e https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:437
    c https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    _invoke https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    t https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:857
    g https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    promise callback*g https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    promise callback*g https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    a https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    f https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    f https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:416
    p https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:437
    onRouteChange https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2787
    Qe https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    $emit https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:27
    n https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2850
    beforeMount https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2850
    p https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    r https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    r https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
    pn https://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0:2806
talk.js:2828:6618
10:37:24.548 Current participant not found talk.js:938:1547
10:37:24.553 The request has been canceled 
Object { message: undefined }
talk.js:1008:9499
10:37:25.275 GEThttps://nextcloud.ourdomain.de/index.php/apps/spreed/not-found
[HTTP/2 200 OK 352ms]

10:37:25.658 Navigating away, leaving conversation talk.js:2850:18144
10:37:26.204 GEThttps://nextcloud.ourdomain.de/apps/user_status/img/app.svg?v=81cfd9f1
10:37:26.210 GEThttps://nextcloud.ourdomain.de/core/js/dist/main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.214 GEThttps://nextcloud.ourdomain.de/core/js/dist/files_fileinfo.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.219 GEThttps://nextcloud.ourdomain.de/core/js/dist/files_client.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.240 GEThttps://nextcloud.ourdomain.de/index.php/js/core/merged-template-prepend.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.248 GEThttps://nextcloud.ourdomain.de/apps/files_pdfviewer/js/files_pdfviewer-public.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.250 GEThttps://nextcloud.ourdomain.de/apps/files_sharing/js/dist/main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.252 GEThttps://nextcloud.ourdomain.de/apps/files_videoplayer/js/main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.254 GEThttps://nextcloud.ourdomain.de/apps/notifications/js/notifications-main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.255 GEThttps://nextcloud.ourdomain.de/apps/files_pdfviewer/js/files_pdfviewer-main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.264 GEThttps://nextcloud.ourdomain.de/apps/viewer/js/viewer-main.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.267 GEThttps://nextcloud.ourdomain.de/custom_apps/onlyoffice/js/viewer.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.270 GEThttps://nextcloud.ourdomain.de/custom_apps/onlyoffice/js/listener.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.273 GEThttps://nextcloud.ourdomain.de/apps/files_rightclick/js/script.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.275 GEThttps://nextcloud.ourdomain.de/apps/files_rightclick/js/files.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.280 GEThttps://nextcloud.ourdomain.de/apps/firstrunwizard/js/about.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.282 GEThttps://nextcloud.ourdomain.de/apps/theming/js/theming.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.284 GEThttps://nextcloud.ourdomain.de/apps/user_status/js/user-status-menu.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.286 GEThttps://nextcloud.ourdomain.de/custom_apps/spreed/js/talk.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.288 GEThttps://nextcloud.ourdomain.de/core/js/dist/unified-search.js?v=81cfd9f1-0
[HTTP/2 200 OK 0ms]

10:37:26.790 No OC found index.js:46:12
10:37:26.833 JQMIGRATE: Migrate is installed, version 1.4.1 jquery-migrate.min.js:2:551
10:37:26.885 jQuery is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. globals.js:60:15
10:37:26.886 $ is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. 4 globals.js:60:15
10:37:26.952 Notifications permissions granted App.vue:296
10:37:27.062 GEThttps://nextcloud.ourdomain.de/apps/notifications/img/notifications.svg

10:37:27.064 OCA.Viewer initialized Viewer.js:46:10
10:37:27.067 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/notifications/api/v2/notifications
[HTTP/2 200 OK 320ms]

10:37:27.068 GEThttps://nextcloud.ourdomain.de/apps/notifications/img/notifications.svg
10:37:27.072 No OCA.Files app found, viewer is now in standalone mode Viewer.vue:298
10:37:27.074 $ is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. globals.js:60:15
10:37:27.079 jQuery is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. 2 globals.js:60:15
10:37:27.206 $ is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. globals.js:60:15
10:37:27.408 jQuery is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. globals.js:60:15
10:37:27.499 XHRPUThttps://nextcloud.ourdomain.de/index.php/apps/user_status/heartbeat
[HTTP/2 204 No Content 344ms]

10:37:27.503 XHRGEThttps://nextcloud.ourdomain.de/index.php/apps/onlyoffice/ajax/settings
[HTTP/2 200 OK 351ms]

10:37:27.794 Setting current user App.vue:329
10:37:27.866 Detected browser Firefox 82 (82.0) browserCheck.js:29:11
10:37:28.211 session heartbeat polling started session-heartbeat.js:97:9
10:37:28.221 Files_PDFViewer initialized for public page 
Object { isPublicPage: null, canDownload: true, isSecureViewerAvailable: false }
public.js:30:9
10:37:28.223 Please do NOT wait for the DOMContentLoaded before registering your viewer handler Viewer.vue:217
10:37:28.382 $ is deprecated: The global jQuery is deprecated. It will be updated to v3.x in Nextcloud 21. In later versions of Nextcloud it might be removed completely. Please ship your own. 2 globals.js:60:15
10:37:28.397 Please do NOT wait for the DOMContentLoaded before registering your viewer handler Viewer.vue:217
10:37:28.398 The following mime is already registered application/pdf 
Object { id: Getter & Setter, group: Getter & Setter, mimes: Getter & Setter, component: Getter & Setter, … }
Viewer.vue:511
10:37:29.529 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v2/room
[HTTP/2 200 OK 346ms]

10:37:29.536 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/search/providers?from=%2Fapps%2Fspreed%2Fnot-found
[HTTP/2 200 OK 321ms]

10:37:29.702 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/user_status
[HTTP/2 200 OK 323ms]

10:37:29.778 GEThttps://nextcloud.ourdomain.de/custom_apps/spreed/img/favicon-touch.png
10:37:29.779 GEThttps://nextcloud.ourdomain.de/custom_apps/spreed/img/favicon.ico
10:37:30.037 [DEBUG] unified-search: Unified Search initialized with the following providers 
Object { 0: {…}, 1: {…}, 2: {…}, 3: {…}, 4: {…}, 5: {…}, 6: {…}, 7: {…}, 8: {…}, 9: {…}, … }
ConsoleLogger.js:42:18
10:37:30.055 X-Nextcloud-Talk-Hash initialised:  d25b6d1b2e13909a0fd3f03a4523abdf3e362b71 talkHashStore.js:64:11
10:37:30.399 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user5/44
10:37:30.409 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user3/44
[HTTP/2 200 OK 0ms]

10:37:30.874 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user1/44

10:37:30.881 GEThttps://nextcloud.ourdomain.de/index.php/avatar/user2/44

10:37:30.905 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/statuses/user5
[HTTP/2 200 OK 421ms]

10:37:30.941 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/statuses/user3
[HTTP/2 200 OK 420ms]

10:37:30.952 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/statuses/user1
[HTTP/2 200 OK 420ms]

10:37:30.955 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/user_status/api/v1/statuses/user2
[HTTP/2 200 OK 420ms]
```
</details>

## Server configuration

**Operating system**: Ubuntu host, nextcloud docker (fpm-alpine)

**Web server:** Caddy V2

**Database:** PostgreSQL

**PHP version:** 7.4

**Nextcloud Version:** 20.0.2

**List of activated apps:**

<details>

**Enabled:**
  - cloud_federation_api: 1.3.0
  - comments: 1.10.0
  - contacts: 3.4.2
  - contactsinteraction: 1.1.0
  - dashboard: 7.0.0
  - dav: 1.16.1
  - federatedfilesharing: 1.10.1
  - files: 1.15.0
  - files_markdown: 2.3.1
  - files_pdfviewer: 2.0.1
  - files_rightclick: 0.17.0
  - files_sharing: 1.12.0
  - files_texteditor: 2.14.0
  - files_trashbin: 1.10.1
  - files_versions: 1.13.0
  - files_videoplayer: 1.9.0
  - firstrunwizard: 2.9.0
  - groupfolders: 8.1.1
  - logreader: 2.5.0
  - lookup_server_connector: 1.8.0
  - mail: 1.7.0
  - nextcloud_announcements: 1.9.0
  - notifications: 2.8.0
  - oauth2: 1.8.0
  - onlyoffice: 6.1.0
  - password_policy: 1.10.1
  - photos: 1.2.0
  - polls: 1.6.2
  - privacy: 1.4.0
  - provisioning_api: 1.10.0
  - recommendations: 0.8.0
  - serverinfo: 1.10.0
  - settings: 1.2.0
  - sharebymail: 1.10.0
  - spreed: 10.0.3
  - support: 1.3.0
  - survey_client: 1.8.0
  - systemtags: 1.10.0
  - theming: 1.11.0
  - twofactor_backupcodes: 1.9.0
  - updatenotification: 1.10.0
  - user_status: 1.0.1
  - viewer: 1.4.0
  - weather_status: 1.0.0
  - workflowengine: 2.2.0

**Disabled:**
  - accessibility
  - activity
  - admin_audit
  - bbb
  - calendar
  - circles
  - deck
  - encryption
  - federation
  - files_external
  - notes
  - richdocuments
  - tasks
  - text
  - user_ldap
  - w2g2
</details>

**Nextcloud configuration:**

<details>

```
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "overwritehost": "nextcloud.ourdomain.de",
        "overwriteprotocol": "https",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": 465,
        "mail_smtpsecure": "ssl",
        "mail_smtpauth": true,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "nextcloud.ourdomain.de",
            "ourdomain.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "20.0.2.2",
        "overwrite.cli.url": "https:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "loglevel": 1,
        "app_install_overwrite": [
            "onlyoffice",
            "w2g2",
            "spreed"
        ],
        "allow_local_remote_servers": true,
        "theme": ""
    }
}
```
</details>

### Server log (data/nextcloud.log)

<details>

Nothing above debug...

```
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"bbb","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"contacts","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"groupfolders","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"files_markdown","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"files_sharing","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"files_texteditor","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"onlyoffice","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"w2g2","method":"GET","url":"/index.php/apps/spreed/not-found","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
{"reqId":"tSAlunkOpFvzo9LCKjOm","level":0,"time":"2020-11-25T08:01:54+00:00","remoteAddr":"37.49.71.43","user":"user4","app":"scss_cacher","method":"GET","url":"/index.php/apps/spreed/not-found","message":"SCSSCacher::process ordinary check follows","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0","version":"20.0.2.2"}
```
</details>

**Comment by PVince81** [comment]
should already be fixed in Talk 10.0.3, please upgrade

ref: https://github.com/nextcloud/spreed/pull/4593

**Comment by PVince81** [comment]
wait, you have 10.0.3 according to the list, are you sure it's still happening there ?

**Comment by PVince81** [comment]
make sure to shift+refresh the page / conversations in question in case it's a caching issue

please also check if the moment where this happens, whether there was a hash sign `#` in the URL.

if not then this is yet another issue

**Comment by fritz-k** [comment]
> wait, you have 10.0.3 according to the list, are you sure it's still happening there ?

I think I upgraded to 10.0.3 a couple of days ago and the problem persists for my 3 colleagues and me.

> make sure to shift+refresh the page / conversations in question in case it's a caching issue

Shift+refresh didn't help, the issue still pops up occasionally.
 
> please also check if the moment where this happens, whether there was a hash sign `#` in the URL.

I didn't see any `#` in the URL.

**Comment by fritz-k** [comment]
@PVince81 I can give you/others a test account on our NC instance and/or the docker and caddy config files if that helps in tracking down the issue.

**Comment by PVince81** [comment]
I'm currently busy with other stuff but will keep an eye. I think I also managed to reproduce this a few times locally.

**Comment by PVince81** [comment]
I haven't seen it happening locally.

Usually this happens when an important server call returned 404, which is interpreted as "not found".

From your log above it seems the signaling server returned 404:
```
10:37:24.061 XHRGEThttps://nextcloud.ourdomain.de/ocs/v2.php/apps/spreed/api/v1/signaling/kjvse3k9
[HTTP/2 404 Not Found 403ms]
```

Next time you see it happening you could check the network console and see if the "signaling/xxxx" call returned 404. "xxxx" is the token from the browser URL. And also whether that call succeeds when there is no "not found" error.

If that is the problem then the question would be why signaling would sometimes return 404...

**Comment by fritz-k** [comment]
I just checked: If `signaling/kjvse3k9` returns a 404 I get redirected to the "not found" page. Otherwise the conversation loads fine.

The requested URL was the same in both cases (although the error triggered 5 times in a row, which is way more than usual for me, but apparently normal for my colleagues).

**Comment by mopsig** [comment]
Hi, I think I am having the same issue.
It only seems to happen to me when I open/switch conversations in the UI, never when accessing the conversation URL directly.
The first call to `/ocs/v2.php/apps/spreed/api/v1/signaling/xxxxxxxx` fails with 404 sometimes, just after opening the chat.

At first I thought this might be related to my configuration having multiple loadbalanced nextcloud instances (nginx) or external php sessions and caching (redis), but it also happens without all that.

**Comment by fritz-k** [comment]
I just checked and can confirm that entering the URL is leading to fewer errors then accessing it via UI. Probably the same observation: Hitting "back" after the error triggered quite reliably opens the conversation on 2nd try.

**Comment by mopsig** [comment]
After enabling session locking as described in https://github.com/phpredis/phpredis#session-locking it now works almost all the time for me.

**Comment by mopsig** [comment]
I now suspect there is a more general configuration issue leading to race conditions.
For example, some users of the mail app get a 404 after submitting their IMAP account info, even though the account gets created and synched.
This makes me doubt that the talk app is to blame here.

**Comment by PVince81** [comment]
In my local setup, when my computer is being slow, if I aggressively switch between two conversations it also ends up in a 404.

**Comment by mkools** [comment]
I have the same issue, it redirects me to: /ocs/v2.php/search/providers?from=%2Fapps%2Fspreed%2Fnot-found

I also can't delete or leave conversations.  Talk just doesn't work. Using latest Nextcloud 20.6 and Talk.

HTTP log shows this
```
 - - [05/Feb/2021:11:20:16 +0000] "GET /apps/spreed/not-found HTTP/2.0" 200 8816 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.146 Safari/537.36"
 - - [05/Feb/2021:11:20:16 +0000] "GET /ocs/v2.php/search/providers?from=%2Fapps%2Fspreed%2Fnot-found HTTP/2.0" 200 272 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.146 Safari/537.36"

```

**Comment by PVince81** [comment]
I'm now able to reproduce this from time to times using a JS script that randomly switches between two conversations while bots are posting messages there:

```javascript
function simulateSwitching(tokens) {
   let newToken = tokens[Math.floor(Math.random() * tokens.length)]
   console.log('Switching to ', newToken)
   document.getElementById('conversation_' + newToken).click()
}

var simulatorInterval = window.setInterval(() => simulateSwitching(['token1', 'token2']), 3000)
```

combine with multiple instances of https://gist.github.com/PVince81/300f03d277a1eb8c876a27b5ddae932e

after a while it triggers the "not found" page.

need to look into the details

**Comment by PVince81** [comment]
note: my test is with internal signaling, not HPB.

with the simulator I was able to capture a 404.

What I see is that during the switch a DELETE is sent to "https://vvortex.local/ocs/v2.php/apps/spreed/api/v3/room/v9gc2hmb/participants/active" and directly after that the signaling call is made to "https://vvortex.local/ocs/v2.php/apps/spreed/api/v2/signaling/v9gc2hmb" and fails with 404, even without a response.

This looks like a race condition.
We should look into cancelling that signaling call or making it fail silently when we already know that the scheduled call is for a room we just left.

**Comment by PVince81** [comment]
As far as I can see and observe, there is already logic for detecting this and cancelling: https://github.com/nextcloud/spreed/blob/master/src/utils/signaling.js#L438

I added more logging locally and indeed I see that this works most of the time.

So there's probably one specific code path that slips through the cracks...

**Comment by PVince81** [comment]
Some things that come to mind:

- [ ] if calling joinRoom(token1) on a room that is pending leaveRoom(token1) leaving due to slowness, then cancel the leaveRoom call? (or provide some kind of mutex)
- [ ] if calling leaveRoom(token1) on a room that is pending joinRoom(token1) leaving due to slowness, then cancel the joinRoom call? (or provide some kind of mutex)

**Comment by PVince81** [comment]
A catch:
<img width="648" alt="image" src="https://user-images.githubusercontent.com/277525/108388693-2bcf1680-720f-11eb-8fa4-2e9cb98cff5d.png">

As we can see, when `leaveRoom` is called first with the token, there's some async stuff happening until it can say "ok, we're done leaving and clearing the currentRoomToken": https://github.com/nextcloud/spreed/blob/master/src/utils/signaling.js#L224

And also we can see that `joinRoom` hasn't fully started yet for the new room, so `startPullingMessages` is still running for the old room that we are pending to leave. 

And then leaveConversation does the DELETE to "/chat/{token}/active" so the server considers that we left already, but signaling frontend isn't up to date yet. So when `startPullingMessages` receives a 404 error, the `this.currentRoomToken` is still equal to the old value here https://github.com/nextcloud/spreed/blob/master/src/utils/signaling.js#L438. So it doesn't know that it should ignore the 404.

The `this.currentRoomToken` is only updated at the very end of the `signaling.leaveRoom()` call promises chain.

Some ideas to fix this:

1) When `leaveRoom` is called, directly clear `this.currentRoomToken`. Not sure if this could cause side effects. And what to do in case of errors ?
or
2) When `leaveRoom` is called, set a flag to say "we're pending leaving" and then the other calls like `startPullingMessages` know that we're about to leave and can skip the 404 properly.
or
3) When `leaveRoom` is called, cancel any pending `startPullingMessages`. Might not work as subsequent calls might happen.

For now I'd tend to solution 2.
And add a similar thing for `joinRoom` so we can know when we're "pending join".

**Comment by PVince81** [comment]
4. check if the token in the URL is different than the one from the 404, then assume we already switched to the new room ? as we're not sitting in a potentially non-existing room we can assume that the signaling's 404 is not about "room not found" but "we left already"

**Comment by PVince81** [comment]
The latter could be achieved by adding a "token" argument to the new "not found" event in https://github.com/nextcloud/spreed/pull/5082 (pending review/merge). Then the event handler can compare the given token to whatever is in the store to see if the "not found" is actually about the currently displayed conversation or another one that we already left. If current: send to "not found", otherwise ignore.

**Comment by PVince81** [comment]
Seems that this PR will bring in additional attributes, we might be able to use them: https://github.com/nextcloud/spreed/pull/5084/files#diff-fd8947d9afe55478ba8e2e1aaec637de34304bc109cd317b5adcece86cbd815eR33

**Comment by raimund-schluessler** [comment]
I also have this problem on my instance. Switching conversations triggers the error like 80 % of the time. Let me know, if I can somehow help debugging this.

**Comment by mkools** [comment]
I get this 100% of the time. Talk is unusable.

**Comment by PVince81** [comment]
The issue I found is when switching conversations too quickly. Basically, when conversation A was not fully loaded, already switch to conversation B.

@mkools if you have it 100% of the time and it happens without switching then it's likely a different issue and not the bug in question. Maybe a setup issue ?

**Comment by raimund-schluessler** [comment]
For me it not only happens when switching conversations. It also happens directly after loading Talk. See this screencapture:

![Talk-Not-Found](https://user-images.githubusercontent.com/2496460/110086748-0b689580-7d93-11eb-9e3f-51d7eae131d1.gif)

I run Nextcloud using the provided Docker containers (Nextcloud, Redis, MariaDB, cron job and a container for notify_push behind a traefik2 proxy). I would be surprised if it was a setup issue, since I basically use the default configuration and the requests reach the Nextcloud instance (so I don't think it's a configuration issue of traefik2). If wanted, I can provide server and client logs for this issue in the evening.

**Comment by mkools** [comment]
> The issue I found is when switching conversations too quickly. Basically, when conversation A was not fully loaded, already switch to conversation B.
> 
> @mkools if you have it 100% of the time and it happens without switching then it's likely a different issue and not the bug in question. Maybe a setup issue ?

Not sure. I'm using nginx though but with the rules provided by Nextcloud. I run on Debian 10 with a haproxy setup. I already tried a fresh Ubuntu 20.04 install with nginx as well but without haproxy and it's still the same issue. I haven't tried Apache yet.

I do run it on a NAT/internal IP VPS, could that maybe be the issue? I did install the complete high performance backend though which all shows ok in the configuration.

**Comment by PVince81** [comment]
If you get a 404 when switching quickly, I already have ways to reproduce it by having a script switch very quickly between rooms repeatedly: https://github.com/nextcloud/spreed/issues/4670#issuecomment-780694721

Try switching more slowly after waiting for full conversation load. The switching should then work correctly.

So I'd say "if switching quickly happens for you, it's a known issue". If joining a room initially without switching gives 404, it's either another bug or an env issue.

**Comment by PVince81** [comment]
I've just tried locally with approach 4 https://github.com/nextcloud/spreed/issues/4670#issuecomment-781475812 and it doesn't fully work in my simulation. It is still possible to trigger a 404 by switching from A to B and quickly back to A. So the URL is not a good source.

**Comment by PVince81** [comment]
I've pushed my attempt to https://github.com/nextcloud/spreed/pull/5322 but will require more work or a different approach to also catch the A->B->A switching cases.

**Comment by PVince81** [comment]
I've updated the subject.

From now on this ticket is for those with issues when switching conversations, not initial join.

For those with problems with initial join (not switching), please move to this other ticket: https://github.com/nextcloud/spreed/issues/5148

This will avoid mixing up logs and test results.

**Comment by PVince81** [comment]
Tried approach 2:

> When leaveRoom is called, set a flag to say "we're pending leaving" and then the other calls like startPullingMessages know that we're about to leave and can skip the 404 properly.

I did so by adding a structure `this._leavingRoom[token] = true` which I clear at the very end of leaveRoom's async process.
Still, as far as I can see `pullMessages` can still have a pending slow call that returns 404 even after that.

Next up: new approach 5:

5. Cancel the pullMessagesRequest directly when leaveRoom is called.

That ought to work!

**Comment by PVince81** [comment]
I've tried the cancel request approach but it only seems to avoid the error if I cancel it regardless of the token.
There's a risk that we might be cancelling the messages of the newly joined room instead of the ones from the one we're leaving.

Now if I'm only cancelling the pullMessagesRequest for the token passed to leaveRoom, it still fails in this crazy quickly swapping scenario:
<img width="387" alt="image" src="https://user-images.githubusercontent.com/277525/110146670-1f37ea00-7ddb-11eb-9220-0a3db247402e.png">

The switching is so fast that we have two `joinRoom` requests that finish for the same room, before the `leaveRoom` for that same token has even finished.

Cancelling the join or leave requests would be risky as we don't know if the server has already acknowledged or not.

So probably we'll rather need to add some client-side locking in place: only allow a maximum of 1 request for leaving and 1 request for joining. If more is incoming, then wait for the previous ones to finish. That might or might not cause some delay in the UI but only when switching too quickly.

**Comment by PVince81** [comment]
A few ideas to explore:
1. Make `joinRoom` and `leaveRoom` requests cancellable, by token. So if the method is called twice for the same token, cancel the request and redo it.

2. Same as 1 but instead of cancelling, just ignore.

3. Extras to 1 and 2: if calling `joinRoom` on a room where `leaveRoom` is in progress, cancel the latter ? This is where it could get thorny.

Extra concern: there's a timeout at the end of "_startPullingMessages()". What if we just finished switching while that one calls `_startPullingMessages()` again while we were switching ? I guess it will simply use `this.currentRoomToken` so if the token changed it would just use that. Should be safe I guess...

**Comment by PVince81** [comment]
Note: my simulator only switches between two rooms currently.

I added more precision in the messages, there's a little mystery here: even though we clicked on "vkdo5v6k", after a few rows `joinRoom` is still called with the other room's token. This can be seen in the blue circles.

![image](https://user-images.githubusercontent.com/277525/110151372-b0f62600-7de0-11eb-9785-1ec7d53264a3.png)

and somehow that messes up the `leaveRoom` call later where while we want to leave "vkdo5v6k" the current room is actually already the other one, see blue circles...

I suspect that the bogus `joinRoom` is probably lagging behind an older switch.

So this still confirms that we should do something to prevent or cancel subsequent join/leave calls for the same rooms.

**Comment by PVince81** [comment]
I've pushed the code that generates those messages here: 43b533e9580a298d5d67492fc4c53ea7422349f7

**Comment by mgriego** [comment]
If you're using the standard Nextcloud Docker containers with Redis, then you definitely have an issue with session locking, as the Redis session handler does NOT lock sessions by default.  This causes a race condition seen here.  A lot of the code here is predicated on the sessions being locked.  I have yet to follow up on this PR, but if you add these settings shown that PR (https://github.com/nextcloud/docker/pull/1364/files), if will enable session locking in the Redis session handler and prevent the race condition you're seeing.  These settings can be added to a local .conf file that you copy into your Nextcloud image in your Dockerfile:

```
redis.session.locking_enabled = "
redis.session.lock_retries = -1
redis.session.lock_wait_time = 10000
```

**Comment by simonsan** [comment]
Is there any way to fix this client-side? A patch for a js library or so?

**Comment by orangesunny** [comment]
I have this issue with the latest Nextcloud and Talk, Docker installation.

When I get https://****/apps/spreed/not-found

Workaround for me so far:
1. Visit another NC app via the top menu, wait until it’s loaded
2. Switch back to Talk
3. Choose the **same** conversation again
4. no https://****/apps/spreed/not-found 🥳 

PITA, but at least usable. 😌

**Comment by dgalarreta** [comment]
Same error hier 

We also changed to Nginx Reverse Proxy but nothing (before it, we had a Traefik proxy Container for it)

No way to use nicely Talk on the Web browser. But with the Android App and iOS App Talk works like a charm!!

We also implemented a High backend for Talk but nothing.

Are you using Mariadb for the NExtcloud Docker Container? Or also the same error with mysql or postgresql

Anyway thanks to the Nextcloud Community!

**Comment by simonsan** [comment]
For us it started with an update to 20.0.8 (I think), I upgraded yesterday to 21.0.0 to see if it fixed, but nope. :(
The cloud is running more slowly than ever and Talk is basically unusable in the browser. :/

**Comment by orangesunny** [comment]
I am anxious to jinx it, but it seems fixed at the moment.
🤞🏼 
Docker installation, NC 21.01, Talk 11.1.12.

**Comment by raimund-schluessler** [comment]
I can also confirm that it doesn't happen anymore with the Docker container 21.0.1 which includes a settings fix for Redis.

**Comment by nickvergessen** [comment]
So lets close it for now

**Comment by fritz-k** [comment]
The issue still persists. I can reliably reproduce it by clicking through 2-3 different conversations before the current one finished loading. Admittedly, this triggers quite infrequently in regular use, but still occasionally when I open the wrong conversation and quickly switch to the correct one.

**Comment by nickvergessen** [comment]
skjnldsv also reported it today on 22, so reopening :(

**Comment by nickvergessen** [comment]
Useful foot print:
![20210707_081630](https://user-images.githubusercontent.com/213943/124709961-686c8780-defc-11eb-934a-599a809b56de.png)


So the problem is you need a slow responding server, so the signaling killing arrives after you started being in the new room.

**Comment by pcevela** [comment]
Well, I can replicate it also on 23.0.1. Is it known what is causing this?

**Comment by DavidMMad** [comment]
I also have the same problem with 22.2.3

**Comment by grickard** [comment]
I have the same thing in 23.0.1. It seems to happen intermittently once a NC login session has been started. It always happens when opening Talk for the first time in a new session. I have multiple instances running behind HAProxy and all utilizing redis.

Here's my console output:

`exports createError.js:16

onabort xhr.js:74

exports xhr.js:65

exports xhr.js:13

exports dispatchRequest.js:50

promise callback\*30321/l.prototype.request Axios.js:61

t Axios.js:76

exports bind.js:9

q messagesService.js:59

request cancelableRequest.js:50

lookForNewMessages messagesStore.js:688

v vuex.esm.js:736

dispatch vuex.esm.js:490

dispatch vuex.esm.js:393

getNewMessages MessagesList.vue:511

lookForNewMessages MessagesList.vue:487

handleStartGettingMessagesPreconditions MessagesList.vue:448

handler MessagesList.vue:232

VueJS 28


MessagesList.vue:548`

**Comment by timm2k** [comment]
Hi,

if you are using Redis session handler please check your php-redis config:
```
redis.session.locking_enabled=1
redis.session.lock_retries=-1
redis.session.lock_wait_time=10000
```

This solved the not-found redirect for me. This config is documented at https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#id2 and in #5148.

Kind regards
Timm

**Comment by nickvergessen** [comment]
This is now happening more and more.
Not sure why

**Comment by andyxheli** [comment]
For me I don't have an issue on 24.0.5 but on NC 25 Beta 7 I am having this issue.

**Comment by andyxheli** [comment]
Same results on NC 25 RC2

**Comment by DrZoidberg09** [comment]
After changing redis settings as stated above it seems to be more stable, but this exact problem persists.

**Comment by nickvergessen** [comment]
After some more analysis I found the root cause of the issue.
The leaveRoom of the old conversation and the joinRoom of the new conversation read the session in parallel modify their data and then save it again. The problem is that both can overwrite each others data and we actually might need to store a key per conversation instead of an array with key "sessions", so parallel stuff is no issue.

**Comment by pjrobertson** [comment]
Just an update with Nextcloud 28 - I am still running into this issue. I edited the php.ini settings as per the post [here](https://github.com/nextcloud/spreed/issues/4670#issuecomment-1232615657) (now also in the official docs [here](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#using-the-redis-session-handler)) and the "not found" errors are less frequest, but they still occur if I switch between chats. 

I am on a slow internet connection, which means loading each chat takes 2-3s – I think this may be aggravating the race condition/locking issue that still seems to be present.

Happy to provide more debug info / open a new issue if you point me towards what is needed – I could do a video but do not want to share sensitive internal data.

**Comment by pschichtel** [comment]
@pjrobertson do you still have this issue and can you confirm if this happens after session.gc_maxlifetime expired? I just reported #12449 for that which might be the same issue you are having.

**Comment by pjrobertson** [comment]
The problem has been fixed there, I can't remember which version, but it
was just before NC 29 was released. Thanks!
